# Preprocessing


In [5]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

In [6]:
# Import and read the american_bankruptcy.csv.

url = "https://raw.githubusercontent.com/jannatulhashi/American_Company_Bankruptcy_Prediction/main/american_bankruptcy.csv"
bankruptcy_df = pd.read_csv(url)
bankruptcy_df.head()


company_name status_label  year  current_assets(X1)  cost_of_goods_sold(X2)  \
0          C_1        alive  1999             511.267                 833.107   
1          C_1        alive  2000             485.856                 713.811   
2          C_1        alive  2001             436.656                 526.477   
3          C_1        alive  2002             396.412                 496.747   
4          C_1        alive  2003             432.204                 523.302   

   depreciation_and_amortization(X3)  EBITDA(X4)  inventory(X5)  \
0                             18.373      89.031        336.018   
1                             18.577      64.367        320.590   
2                             22.496      27.207        286.588   
3                             27.172      30.745        259.954   
4                             26.680      47.491        247.245   

   net_income(X6)  total_receivables(X7)  ...  net_sales(X9)  \
0          35.163                128.348  ...       1024.333   
1          18.531                115.187  ...        874.255   
2         -58.939                 77.528  ...        638.721   
3         -12.410                 66.322  ...        606.337   
4           3.504                104.661  ...        651.958   

   total_assets(X10)  total_long_term_debt(X11)  EBIT(X12)  gross_profit(X13)  \
0            740.998                    180.447     70.658            191.226   
1            701.854                    179.987     45.790            160.444   
2            710.199                    217.699      4.711            112.244   
3            686.621                    164.658      3.573            109.590   
4            709.292                    248.666     20.811            128.656   

   total_current_liabilities(X14)  retained_earnings(X15)  total_revenue(X16)  \
0                         163.816                 201.026            1024.333   
1                         125.392                 204.065             874.255   
2                         150.464                 139.603             638.721   
3                         203.575                 124.106             606.337   
4                         131.261                 131.884             651.958   

   total_liabilities(X17)  total_operating_expenses(X18)  
0                 401.483                        935.302  
1                 361.642                        809.888  
2                 399.964                        611.514  
3                 391.633                        575.592  
4                 407.608                        604.467  

[5 rows x 21 columns]

### Variable Name Description

**X1 :** Current assets - All the assets of a company that are expected to be sold or used as a result of standard business operations over the next year

**X2 :** Cost of goods sold - The total amount a company paid as a cost directly related to the sale of products

**X3 :** Depreciation and amortization - Depreciation refers to the loss of value of a tangible fixed asset over
time (such as property, machinery, buildings, and plant). Amortization refers to the loss of value of intangible assets over time.

**X4 :** EBITDA - Earnings before interest, taxes, depreciation, and amortization. It is a measure of a company's overall financial performance, serving as an alternative to net income.

**X5 :** Inventory - The accounting of items and raw materials that a company either uses in production or sells.

**X6 :** Net Income - The overall profitability of a company after all expenses and costs have been deducted from total revenue.

**X7 :** Total Receivables - The balance of money due to a firm for goods or services delivered or used but not yet paid for by customers.

**X8 :** Market value - The price of an asset in a marketplace. In this dataset, it refers to the market capitalization since companies are publicly traded in the stock market.

**X9 :** Net sales - The sum of a company's gross sales minus its returns, allowances, and discounts.

**X10 :** Total assets - All the assets, or items of value, a business owns.

**X11 :** Total Long-term debt - A company's loans and other liabilities that will not become due within one year of the balance sheet date.

**X12 :** EBIT - Earnings before interest and taxes.

**X13 :** Gross Profit - The profit a business makes after subtracting all the costs that are related to manufacturing and selling its products or services.

**X14 :** Total Current Liabilities - The sum of accounts payable, accrued liabilities, and taxes such as Bonds payable at the end of the year, salaries, and commissions remaining.

**X15 :** Retained Earnings - The amount of profit a company has left over after paying all its direct costs, indirect costs, income taxes, and its dividends to shareholders.

**X16 :** Total Revenue - The amount of income that a business has made from all sales before subtracting expenses. It may include interest and dividends from investments.

**X17 :** Total Liabilities - The combined debts and obligations that the company owes to outside parties.

**X18 :** Total Operating Expenses - The expenses a business incurs through its normal business operations.

In [7]:
bankruptcy_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78682 entries, 0 to 78681
Data columns (total 21 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   company_name                       78682 non-null  object 
 1   status_label                       78682 non-null  object 
 2   year                               78682 non-null  int64  
 3   current_assets(X1)                 78682 non-null  float64
 4   cost_of_goods_sold(X2)             78682 non-null  float64
 5   depreciation_and_amortization(X3)  78682 non-null  float64
 6   EBITDA(X4)                         78682 non-null  float64
 7   inventory(X5)                      78682 non-null  float64
 8   net_income(X6)                     78682 non-null  float64
 9   total_receivables(X7)              78682 non-null  float64
 10  market_value(X8)                   78682 non-null  float64
 11  net_sales(X9)                      78682 non-null  flo

In [8]:
# Check missing values

bankruptcy_df.isnull().sum()

company_name                         0
status_label                         0
year                                 0
current_assets(X1)                   0
cost_of_goods_sold(X2)               0
depreciation_and_amortization(X3)    0
EBITDA(X4)                           0
inventory(X5)                        0
net_income(X6)                       0
total_receivables(X7)                0
market_value(X8)                     0
net_sales(X9)                        0
total_assets(X10)                    0
total_long_term_debt(X11)            0
EBIT(X12)                            0
gross_profit(X13)                    0
total_current_liabilities(X14)       0
retained_earnings(X15)               0
total_revenue(X16)                   0
total_liabilities(X17)               0
total_operating_expenses(X18)        0
dtype: int64

In [9]:
# Determine the number of unique values in each column.
bankruptcy_df.nunique()

company_name                          8971
status_label                             2
year                                    20
current_assets(X1)                   65895
cost_of_goods_sold(X2)               65690
depreciation_and_amortization(X3)    36010
EBITDA(X4)                           59060
inventory(X5)                        38898
net_income(X6)                       55550
total_receivables(X7)                49577
market_value(X8)                     77580
net_sales(X9)                        68596
total_assets(X10)                    71521
total_long_term_debt(X11)            39741
EBIT(X12)                            56949
gross_profit(X13)                    64952
total_current_liabilities(X14)       58685
retained_earnings(X15)               72062
total_revenue(X16)                   68596
total_liabilities(X17)               64640
total_operating_expenses(X18)        70840
dtype: int64

In [10]:
# Look at company_name value counts for binning

company_name_counts = bankruptcy_df["company_name"].value_counts()
company_name_counts

C_4081    20
C_6487    20
C_1153    20
C_4047    20
C_4056    20
          ..
C_1498     1
C_5908     1
C_2814     1
C_3997     1
C_3408     1
Name: company_name, Length: 8971, dtype: int64

In [11]:
# Choose a cutoff value and create a list of company_name to be replaced and  use the variable name `company_name_to_replace`

cutoff_value = 100
company_name_to_replace = company_name_counts[company_name_counts < cutoff_value].index.tolist()

# Replace in dataframe
for name in company_name_to_replace:
    bankruptcy_df['company_name'] =  bankruptcy_df['company_name'].replace(name, "other")

# Check to make sure binning was successful
bankruptcy_df['company_name'].value_counts()

other    78682
Name: company_name, dtype: int64

In [12]:
# Apply one-hot encoding to the `company_name`

bankruptcy_df = pd.get_dummies(bankruptcy_df, columns=['company_name'])


In [13]:
# Convert the status_label column to binary numeric values, with 'alive' represented as 1 and 'failed' as 0
bankruptcy_df['status_label'] = bankruptcy_df['status_label'].apply(lambda x: 1 if x == 'alive' else 0)


In [64]:
bankruptcy_df.tail(20)

status_label  year  current_assets(X1)  cost_of_goods_sold(X2)  \
78662             1  2014              58.314                   1.833   
78663             1  2015              21.381                   4.972   
78664             1  2016               9.049                   1.077   
78665             1  2017               3.369                   0.208   
78666             1  2014            1578.400                 324.400   
78667             1  2015            2270.500                1216.700   
78668             1  2016            2071.200                1723.300   
78669             1  2017            2340.600                1833.200   
78670             1  2018            2482.200                 966.700   
78671             1  2013             923.200                1206.000   
78672             1  2014             973.800                1185.500   
78673             1  2015             735.100                1132.900   
78674             1  2016             829.300                1190.400   
78675             1  2017            1032.700                1269.600   
78676             1  2018             931.600                1475.900   
78677             1  2014             233.211                  43.338   
78678             1  2015             105.559                  59.184   
78679             1  2016              63.971                  69.074   
78680             1  2017             135.207                  66.527   
78681             1  2018              82.589                  68.817   

       depreciation_and_amortization(X3)  EBITDA(X4)  inventory(X5)  \
78662                              0.656     -45.934          5.488   
78663                              0.486     -33.631          1.025   
78664                              0.190     -12.783          0.213   
78665                              0.057     -10.647          0.000   
78666                             86.700     211.600        205.800   
78667                            250.900     524.000        517.500   
78668                            342.300     723.600        416.400   
78669                            354.200     758.100        490.400   
78670                            156.600     412.700        188.100   
78671                             45.600     409.000        314.600   
78672                             44.000     439.300        360.600   
78673                             48.300     444.100        204.100   
78674                             61.400     490.000        220.600   
78675                             62.100     567.300        239.800   
78676                             82.500     619.000        280.300   
78677                             14.094      45.615          3.376   
78678                             42.592     202.133          2.288   
78679                             65.057      79.051          2.581   
78680                             65.330      69.171          2.013   
78681                             65.201      67.262          2.112   

       net_income(X6)  total_receivables(X7)  market_value(X8)  ...  \
78662         -48.205                  0.470           17.0644  ...   
78663         -35.161                  0.040            0.1996  ...   
78664         -13.116                  0.030            0.1309  ...   
78665         -10.890                  0.040            0.2231  ...   
78666         -29.900                327.300         4227.5957  ...   
78667        -308.600               1148.000         2944.0231  ...   
78668         -73.700               1054.800         2788.2080  ...   
78669        -296.200               1156.000         2850.9981  ...   
78670        -324.400                382.400         2985.1014  ...   
78671          31.000                266.100         4243.5215  ...   
78672         175.200                259.900         5314.7873  ...   
78673         153.900                303.400         6327.7267  ...   
78674         229.100                271.900         6097.

Here,
'company_name' represents the name of the company. We've applied binning and one-hot encoding to this column to reduce the number of unique values and transform it into a format suitable for modeling.

'status_label', the target variable of the data set.

'year', the column to split the data into training, test and verification sets.

'X1' - 'X18', the 18 predictor variables.

In [15]:
# Split our preprocessed data into our features and target arrays
X = bankruptcy_df.drop(columns='status_label', axis=1)
y = bankruptcy_df["status_label"]

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [16]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Compile, Train and Evaluate the Model

In [17]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

# Create a sequential model
nn = tf.keras.models.Sequential()

# Input Layer
nn.add(tf.keras.layers.Dense(units=128, activation='relu', input_dim=X_train_scaled.shape[1]))

# First Hidden Layer
nn.add(tf.keras.layers.Dense(units=64, activation='relu'))

# Second Hidden Layer
nn.add(tf.keras.layers.Dense(units=32, activation='relu'))

# Third Hidden layer
nn.add(tf.keras.layers.Dense(units=16, activation='relu'))

# Output Layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 128)               2688      
                                                                 
 dense_1 (Dense)             (None, 64)                8256      
                                                                 
 dense_2 (Dense)             (None, 32)                2080      
                                                                 
 dense_3 (Dense)             (None, 16)                528       
                                                                 
 dense_4 (Dense)             (None, 1)                 17        
                                                                 
Total params: 13569 (53.00 KB)
Trainable params: 13569 (53.00 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [18]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [19]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,validation_split=0.15, epochs=100)

Epoch 1/100
1672/1672 [==============================] - 5s 2ms/step - loss: 0.2376 - accuracy: 0.9334 - val_loss: 0.2300 - val_accuracy: 0.9324
Epoch 2/100
1672/1672 [==============================] - 5s 3ms/step - loss: 0.2256 - accuracy: 0.9343 - val_loss: 0.2308 - val_accuracy: 0.9324
Epoch 3/100
1672/1672 [==============================] - 4s 3ms/step - loss: 0.2230 - accuracy: 0.9343 - val_loss: 0.2252 - val_accuracy: 0.9324
Epoch 4/100
1672/1672 [==============================] - 4s 2ms/step - loss: 0.2207 - accuracy: 0.9347 - val_loss: 0.2241 - val_accuracy: 0.9332
Epoch 5/100
1672/1672 [==============================] - 4s 3ms/step - loss: 0.2183 - accuracy: 0.9353 - val_loss: 0.2201 - val_accuracy: 0.9335
Epoch 6/100
1672/1672 [==============================] - 4s 2ms/step - loss: 0.2166 - accuracy: 0.9350 - val_loss: 0.2270 - val_accuracy: 0.9324
Epoch 7/100
1672/1672 [==============================] - 4s 3ms/step - loss: 0.2154 - accuracy: 0.9355 - val_loss: 0.2204 - val_ac

In [20]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

492/492 - 1s - loss: 0.2355 - accuracy: 0.9356 - 640ms/epoch - 1ms/step
Loss: 0.23546384274959564, Accuracy: 0.9355658888816833


In [21]:
# Save and export the results to an HDF5 file
# Export model to HDF5 File

from google.colab import files

nn.save('/content/american_company_bankruptcy.h5')
files.download('/content/american_company_bankruptcy.h5')


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3000: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# American Company Bankruptcy Analysis:

**Overview of the Analysis:**

The primary objective of this analysis is to predict future bankruptcy events of companies using various financial metrics and attributes. By leveraging deep learning techniques, we aim to develop a model that can accurately identify companies at risk of bankruptcy, allowing stakeholders to make informed decisions.

#### Results:
**Data Preprocessing:**

**Target Variable(s) for the Model:**

**status_label:** This binary variable indicates whether a company is 'alive' or 'failed'.

**Features for the Model:**

Financial metrics such as:

- Current assets (X1)

- Cost of goods sold (X2)

- Depreciation and amortization (X3)

- EBITDA (X4)

- Inventory (X5)

- Net Income (X6)

- Total Receivables (X7)

- Market value (X8)

- Net sales (X9)

- Total assets (X10)

- Total Long-term debt (X11)

- EBIT (X12)

- Total Current Liabilities (X13)

- Retained Earnings (X14)

- Total Revenue (X15)

- Total Liabilities (X16)

- Total Operating Expenses (X18).

**One-hot encoded "company_name".**

**Variables to be Removed:**

The original **"company_name"** column, as it was transformed into a one-hot encoded format.

**Compiling, Training, and Evaluating the Model:**

**Neural Network Architecture:**
Number of Layers: 5 (1 Input, 3 Hidden, 1 Output)

**Neurons:**

- Input layer: 128 neurons

- First hidden layer: 64 neurons

- Second hidden layer: 32 neurons

- Third hidden layer: 16 neurons

- Output layer: 1 neuron

**Activation Functions:**
- 'relu' for input and hidden layers.
- 'sigmoid' for the output layer.

**Reason for Architecture Choice:**

The choice of neurons and layers was based on the complexity of the dataset. The 'relu' activation function is commonly used in hidden layers because of its efficiency and non-linearity. The 'sigmoid' activation function is suitable for the binary classification task at hand.

**Model Performance:**

- Loss: 0.2724533975124359
- Accuracy: 93.72%

**Steps to Increase Model Performance:**

The data was scaled using **StandardScaler**.
One-hot encoding was applied to categorical variables.
The dataset was split into training and testing subsets to evaluate model performance on unseen data.

**Summary:**

The deep learning model developed for predicting bankruptcy events has shown promising results, with an accuracy rate of approximately 93.72%. The model utilized various financial metrics as features and underwent preprocessing steps such as scaling and one-hot encoding for categorical variables.

**Recommendations:**

While our deep learning model has shown notable success, introducing ensemble techniques, such as Random Forest, could further enhance its accuracy.Ensemble methods often provide robust predictions by combining multiple models' outputs. Moreover, techniques like feature importance from ensemble methods can offer insights into which financial metrics are most predictive of bankruptcy.

In [22]:
import os
# spark_version = 'spark-3.4.1'
spark_version = 'spark-3.4.1'
os.environ['SPARK_VERSION']=spark_version

# Install Spark and Java
!apt-get update
!apt-get install openjdk-11-jdk-headless -qq > /dev/null
!wget -q http://www.apache.org/dist/spark/$SPARK_VERSION/$SPARK_VERSION-bin-hadoop3.tgz
!tar xf $SPARK_VERSION-bin-hadoop3.tgz
!pip install -q findspark

# Set Environment Variables
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-11-openjdk-amd64"
os.environ["SPARK_HOME"] = f"/content/{spark_version}-bin-hadoop3"

# Start a SparkSession
import findspark
findspark.init()

Hit:1 http://archive.ubuntu.com/ubuntu jammy InRelease
Get:2 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [119 kB]
Get:3 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease [3,626 B]
Get:4 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [109 kB]
Hit:5 https://ppa.launchpadcontent.net/c2d4u.team/c2d4u4.0+/ubuntu jammy InRelease
Hit:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease
Get:7 http://security.ubuntu.com/ubuntu jammy-security InRelease [110 kB]
Hit:8 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease
Hit:9 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease
Hit:10 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease
Get:11 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [1,269 kB]
Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [1,343 kB]
Get:13 http://security.ubuntu.com/ubuntu jammy-securit

After assessing the model's accuracy, PySpark SQL was integrated to delve deeper into the main dataframe **bankruptcy_df**. The use of PySpark provides scalable and efficient data exploration, allowing for a thorough understanding of underlying trends and patterns. This PySpark SQL offered a platform to derive insights from the data, complementing the model's findings and potentially guiding further refinements.

In [23]:
# Import packages
from pyspark.sql import SparkSession
import time

# Create a SparkSession
spark = SparkSession.builder.appName("BankruptcyAnalysis").getOrCreate()
spark_df = spark.createDataFrame(bankruptcy_df)

# Show DataFrame
spark_df.show()


+------------+----+------------------+----------------------+---------------------------------+----------+-------------+--------------+---------------------+----------------+-------------+-----------------+-------------------------+---------+-----------------+------------------------------+----------------------+------------------+----------------------+-----------------------------+------------------+
|status_label|year|current_assets(X1)|cost_of_goods_sold(X2)|depreciation_and_amortization(X3)|EBITDA(X4)|inventory(X5)|net_income(X6)|total_receivables(X7)|market_value(X8)|net_sales(X9)|total_assets(X10)|total_long_term_debt(X11)|EBIT(X12)|gross_profit(X13)|total_current_liabilities(X14)|retained_earnings(X15)|total_revenue(X16)|total_liabilities(X17)|total_operating_expenses(X18)|company_name_other|
+------------+----+------------------+----------------------+---------------------------------+----------+-------------+--------------+---------------------+----------------+-------------+

In [24]:
# Create a temporary view of the DataFrame.

spark_df.createOrReplaceTempView('company_bankruptcy')

In [25]:
# How many companies go bankrupt  each year?

spark.sql("""SELECT year, COUNT(*) as num_bankruptcies_companies
             FROM company_bankruptcy
             WHERE status_label = 0
             GROUP BY year
             ORDER BY year""").show()


+----+--------------------------+
|year|num_bankruptcies_companies|
+----+--------------------------+
|1999|                       380|
|2000|                       404|
|2001|                       414|
|2002|                       414|
|2003|                       415|
|2004|                       404|
|2005|                       379|
|2006|                       366|
|2007|                       336|
|2008|                       284|
|2009|                       234|
|2010|                       218|
|2011|                       194|
|2012|                       182|
|2013|                       167|
|2014|                       142|
|2015|                       111|
|2016|                        83|
|2017|                        57|
|2018|                        36|
+----+--------------------------+



In [26]:
# How many companies remain alive each year?

spark.sql("""SELECT year, COUNT(*) AS num_alive_companies
             FROM company_bankruptcy
             WHERE status_label = 1
             GROUP BY year
             ORDER BY year""").show()


+----+-------------------+
|year|num_alive_companies|
+----+-------------------+
|1999|               4928|
|2000|               4822|
|2001|               4483|
|2002|               4237|
|2003|               4002|
|2004|               3944|
|2005|               3826|
|2006|               3762|
|2007|               3673|
|2008|               3573|
|2009|               3509|
|2010|               3407|
|2011|               3319|
|2012|               3303|
|2013|               3317|
|2014|               3362|
|2015|               3243|
|2016|               3108|
|2017|               2957|
|2018|               2687|
+----+-------------------+



In [27]:
# Which year had the highest number of bankruptcies?

spark.sql("""SELECT year, COUNT(*) as num_bankruptcies
             FROM company_bankruptcy
             WHERE status_label = 0
             GROUP BY year
             ORDER BY num_bankruptcies DESC
             LIMIT 15""").show()

+----+----------------+
|year|num_bankruptcies|
+----+----------------+
|2003|             415|
|2002|             414|
|2001|             414|
|2004|             404|
|2000|             404|
|1999|             380|
|2005|             379|
|2006|             366|
|2007|             336|
|2008|             284|
|2009|             234|
|2010|             218|
|2011|             194|
|2012|             182|
|2013|             167|
+----+----------------+



In [28]:
# What's the average value of each predictor variable (X1 - X18) for "alive" companies?
spark.sql("""SELECT
    AVG(`current_assets(X1)`) as avg_X1,
    AVG(`cost_of_goods_sold(X2)`) as avg_X2,
    AVG(`depreciation_and_amortization(X3)`) as avg_X3,
    AVG(`EBITDA(X4)`) as avg_X4,
    AVG(`inventory(X5)`) as avg_X5,
    AVG(`net_income(X6)`) as avg_X6,
    AVG(`total_receivables(X7)`) as avg_X7,
    AVG(`market_value(X8)`) as avg_X8,
    AVG(`net_sales(X9)`) as avg_X9,
    AVG(`total_assets(X10)`) as avg_X10,
    AVG(`total_long_term_debt(X11)`) as avg_X11,
    AVG(`EBIT(X12)`) as avg_X12,
    AVG(`gross_profit(X13)`) as avg_X13,
    AVG(`total_current_liabilities(X14)`) as avg_X14,
    AVG(`retained_earnings(X15)`) as avg_X15,
    AVG(`total_revenue(X16)`) as avg_X16,
    AVG(`total_liabilities(X17)`) as avg_X17,
    AVG(`total_operating_expenses(X18)`) as avg_X18
FROM company_bankruptcy
WHERE status_label = 1""").show()



+-----------------+------------------+------------------+-----------------+-----------------+------------------+------------------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+-----------------+-----------------+-----------------+------------------+------------------+
|           avg_X1|            avg_X2|            avg_X3|           avg_X4|           avg_X5|            avg_X6|            avg_X7|            avg_X8|           avg_X9|          avg_X10|          avg_X11|          avg_X12|          avg_X13|          avg_X14|          avg_X15|          avg_X16|           avg_X17|           avg_X18|
+-----------------+------------------+------------------+-----------------+-----------------+------------------+------------------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+-----------------+-----------------+-----------------+------------------+------------------+
|

These average values, which cover diverse financial metrics from asset management to profitability indicators, consistently show positive signs. They support the notion that the "alive" companies in this dataset are indeed financially sound, profitable, and efficiently operated.Having a clear grasp of the financial metrics and their implications on a company's health allows for a more informed, accurate, and interpretable bankruptcy prediction model.

In [29]:
# What's the average value of each predictor variable (X1 - X18) for "failed" companies?
spark.sql("""SELECT
    AVG(`current_assets(X1)`) as avg_X1,
    AVG(`cost_of_goods_sold(X2)`) as avg_X2,
    AVG(`depreciation_and_amortization(X3)`) as avg_X3,
    AVG(`EBITDA(X4)`) as avg_X4,
    AVG(`inventory(X5)`) as avg_X5,
    AVG(`net_income(X6)`) as avg_X6,
    AVG(`total_receivables(X7)`) as avg_X7,
    AVG(`market_value(X8)`) as avg_X8,
    AVG(`net_sales(X9)`) as avg_X9,
    AVG(`total_assets(X10)`) as avg_X10,
    AVG(`total_long_term_debt(X11)`) as avg_X11,
    AVG(`EBIT(X12)`) as avg_X12,
    AVG(`gross_profit(X13)`) as avg_X13,
    AVG(`total_current_liabilities(X14)`) as avg_X14,
    AVG(`retained_earnings(X15)`) as avg_X15,
    AVG(`total_revenue(X16)`) as avg_X16,
    AVG(`total_liabilities(X17)`) as avg_X17,
    AVG(`total_operating_expenses(X18)`) as avg_X18
    FROM company_bankruptcy
    WHERE status_label = 0""").show()

+-----------------+----------------+-----------------+-----------------+-----------------+------------------+------------------+-----------------+------------------+------------------+-----------------+-----------------+-----------------+------------------+-------------------+------------------+------------------+------------------+
|           avg_X1|          avg_X2|           avg_X3|           avg_X4|           avg_X5|            avg_X6|            avg_X7|           avg_X8|            avg_X9|           avg_X10|          avg_X11|          avg_X12|          avg_X13|           avg_X14|            avg_X15|           avg_X16|           avg_X17|           avg_X18|
+-----------------+----------------+-----------------+-----------------+-----------------+------------------+------------------+-----------------+------------------+------------------+-----------------+-----------------+-----------------+------------------+-------------------+------------------+------------------+---------------

From the provided data on "failed" companies:

**Operational Losses:** A negative net income (avg_X6) suggests these companies consistently operated at a loss.

**Limited Financial Buffer:** Lower total assets (avg_X10) and a relatively high long-term debt (avg_X11) indicate financial vulnerability and potential liquidity issues.

**Revenue vs. Debt:** Despite generating revenue (avg_X9), these companies had considerable debt (avg_X11 and avg_X14), hinting at possible financial mismanagement or overwhelming expenses.

In summary, "failed" companies in this dataset exhibited consistent operational losses, had limited assets to cover liabilities, and faced significant financial challenges.

In [30]:
import pandas as pd
import numpy as np
from pathlib import Path
import csv
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

import pydotplus
from IPython.display import Image

In [31]:
# Identify the variables

X.head()
X.tail()
y.head()
y.tail()

78677    1
78678    1
78679    1
78680    1
78681    1
Name: status_label, dtype: int64

In [32]:
# Split the data into training and testing

X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    random_state=42,
                                                    stratify=y)

In [33]:
# Identify the Logistic Regression
classifier = LogisticRegression(solver='lbfgs', random_state=1)

classifier

LogisticRegression(random_state=1)

In [61]:
# Train the data
from sklearn.linear_model import LogisticRegression

classifier = LogisticRegression(max_iter=1000)
classifier.fit(X_train, y_train)


LogisticRegression(max_iter=1000)

In [62]:
import warnings
warnings.filterwarnings(action='ignore', category=UserWarning)

# Score the model

# Score the model on the training data
training_score = classifier.score(X_train, y_train)
print(f"Training Data Score: {training_score}")


testing_score = classifier.score(X_test, y_test)
print(f"Testing Data Score: {testing_score}")


Training Data Score: 0.9336564369354866
Testing Data Score: 0.9335570128615729


In [58]:
# Predict outcomes for test data set
predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test})

Prediction  Actual
20172           1       1
29296           1       1
56339           1       1
57739           1       1
41932           1       1
...           ...     ...
57997           1       1
22791           1       1
22119           1       1
11043           1       0
21650           1       1

[19671 rows x 2 columns]

In [59]:
# Calculate the accuracy score
# Display the test score

accuracy_score(y_test, predictions)

0.9335570128615729

# Classification


In [39]:
bankruptcy_df.head()

status_label  year  current_assets(X1)  cost_of_goods_sold(X2)  \
0             1  1999             511.267                 833.107   
1             1  2000             485.856                 713.811   
2             1  2001             436.656                 526.477   
3             1  2002             396.412                 496.747   
4             1  2003             432.204                 523.302   

   depreciation_and_amortization(X3)  EBITDA(X4)  inventory(X5)  \
0                             18.373      89.031        336.018   
1                             18.577      64.367        320.590   
2                             22.496      27.207        286.588   
3                             27.172      30.745        259.954   
4                             26.680      47.491        247.245   

   net_income(X6)  total_receivables(X7)  market_value(X8)  ...  \
0          35.163                128.348          372.7519  ...   
1          18.531                115.187          377.1180  ...   
2         -58.939                 77.528          364.5928  ...   
3         -12.410                 66.322          143.3295  ...   
4           3.504                104.661          308.9071  ...   

   total_assets(X10)  total_long_term_debt(X11)  EBIT(X12)  gross_profit(X13)  \
0            740.998                    180.447     70.658            191.226   
1            701.854                    179.987     45.790            160.444   
2            710.199                    217.699      4.711            112.244   
3            686.621                    164.658      3.573            109.590   
4            709.292                    248.666     20.811            128.656   

   total_current_liabilities(X14)  retained_earnings(X15)  total_revenue(X16)  \
0                         163.816                 201.026            1024.333   
1                         125.392                 204.065             874.255   
2                         150.464                 139.603             638.721   
3                         203.575                 124.106             606.337   
4                         131.261                 131.884             651.958   

   total_liabilities(X17)  total_operating_expenses(X18)  company_name_other  
0                 401.483                        935.302                   1  
1                 361.642                        809.888                   1  
2                 399.964                        611.514                   1  
3                 391.633                        575.592                   1  
4                 407.608                        604.467                   1  

[5 rows x 21 columns]

In [40]:
# Split tbe data into X (features) and y (target)
y = bankruptcy_df["status_label"]
X = bankruptcy_df.drop(columns=["status_label"])

In [41]:
# Split into training and testing set using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [42]:
# Fit a logistic resgression classifier
bankruptcy_logistic_regression_model = LogisticRegression(random_state=6)

In [43]:
# Fit and save logistic regression model using the training data
lgr_model = bankruptcy_logistic_regression_model.fit(X_train, y_train)

In [44]:
# Generate training predictions
training_predictions = lgr_model.predict(X_train)

# Generate testing predictions
testing_predictions = bankruptcy_logistic_regression_model.predict(X_test)

# Confusion Matrix


In [46]:
# Import the model for sklearn confusion matrix
from sklearn.metrics import confusion_matrix

# Predict the labels for the training data
training_predictions = classifier.predict(X_train)

# Create and save the confusion matrix for the training data
training_matrix = confusion_matrix(y_train, training_predictions)

print(training_matrix)




[[   30  3879]
 [   36 55066]]


In [47]:
# Create and save confusion matrix for testing data
test_matrix = confusion_matrix(y_test, testing_predictions)

# Print test matrix
print(test_matrix)


[[   13  1298]
 [    9 18351]]


In [48]:
# Create training classification report
training_report = classification_report(y_train, training_predictions)

print(training_report)

              precision    recall  f1-score   support

           0       0.45      0.01      0.02      3909
           1       0.93      1.00      0.97     55102

    accuracy                           0.93     59011
   macro avg       0.69      0.50      0.49     59011
weighted avg       0.90      0.93      0.90     59011



In [49]:
# Create testing classification report
testing_report = classification_report(y_test, testing_predictions)

# Print the testing classification report
print(testing_report)

              precision    recall  f1-score   support

           0       0.59      0.01      0.02      1311
           1       0.93      1.00      0.97     18360

    accuracy                           0.93     19671
   macro avg       0.76      0.50      0.49     19671
weighted avg       0.91      0.93      0.90     19671



Analysis of Classification

How well does the performance of the  training and test dataset compare?

Answer: When it comes to Active companies, there is a 93% success rate, meanwhile for the failed companies there is only a 59% success rate. There is a lot of unpredictability within the "failed" companies. There could also be external factors as to why a companies goes bankrupt.  

**RANDOM FOREST**